stregsystem/achievements.py

```diff
@@ -0,0 +1,227 @@
+    ]
+
+    return result
+
+
+def get_missing_achievements(member: Member) -> QuerySet[Achievement]:
+    """Gets all missing achievements for a member"""
+    completed_achievements = AchievementComplete.objects.filter(member=member)
+    completed_achievement_ids = completed_achievements.values_list('achievement_id', flat=True)
+    missing_achievements = Achievement.objects.exclude(id__in=completed_achievement_ids)
+
+    return missing_achievements
+
+
+def get_user_leaderboard_position(member: Member) -> float:
+    """
+    Returns the top percentage that the member is in
+    based on number of completed achievements among all users.
+    Users with the same total share the same rank.
+    """
+    # Build leaderboard with total achievement counts
+    leaderboard = (
+        AchievementComplete.objects.all()
+        .values('member')
+        .annotate(total=Count('id'))
+        .order_by('-total', 'member')  # tie-break deterministically
+    )
+
+    if not leaderboard:
+        return 1.0
+
+    # Assign ranks with dense ranking
+    ranks = {}
+    current_rank = 1
+    last_total = None
+
+    for entry in leaderboard:
+        member_id = entry['member']
+        total = entry['total']
+
+        if total != last_total:
+            rank = current_rank
+        # if total == last_total, keep previous rank
+
+        ranks[member_id] = rank
+        last_total = total
+        current_rank += 1
+
+    if member.id not in ranks:
+        return 1.0  # Member has no achievements
+
+    member_rank = ranks[member.id]
+    total_ranks = len(set(ranks.values()))  # total distinct rank positions
+
+    return member_rank / total_ranks
+
+
+def _find_completed_achievements(
+    related_achievements: List[Achievement], member: Member, now: datetime
+) -> List[Achievement]:
+
+    # Filter member's sales to match relevant achievement tasks
+    task_to_sales: Dict[AchievementTask, QuerySet[Sale]] = _filter_relevant_sales(related_achievements, member, now)
+
+    completed_achievements: List[Achievement] = []
+    new_completions: List[AchievementComplete] = []
+
+    for achievement in related_achievements:
+        tasks = achievement.tasks.all()
+
+        if all(task.is_task_completed(task_to_sales[task], member) for task in tasks):
+            completed_achievements.append(achievement)
+            new_completions.append(AchievementComplete(member=member, achievement=achievement))
+
+    if new_completions:
+        AchievementComplete.objects.bulk_create(new_completions)
+
+    return completed_achievements
+
+
+def _filter_relevant_sales(
+    achievements: List[Achievement], member: Member, now: datetime
+) -> Dict[AchievementTask, QuerySet[Sale]]:
+    # Start with all sales for this member, select related to reduce hits
+    member_sales = Sale.objects.filter(member=member).select_related('product').prefetch_related('product__categories')
+    task_to_sales: Dict[int, QuerySet[int]] = {}
```

Comment:
```suggestion
    task_to_sales: Dict[int, QuerySet[int]] = {}
```
Not sure, but maybe?